### PR TITLE
refactor: endpoint resolver exec (SRA)

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

N/A

## Description of changes

* Refactor endpoint resolution to be explicit in `SdkOperationExecution` and change the ordering of endpoint resolution to align with SRA (should be after identity and right before signing)
* Refactor endpoint middleware to be a self contained component that handles all operations. 
* refactored endpoint codegen a bit to align better with recent auth codegen (group endpoint related codegen stuff into a "delegator").

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointDelegator.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    fun generateEndpointParameters(ctx: ProtocolGenerator.GenerationContext, rules: EndpointRuleSet?) {
+        val paramsSymbol = EndpointParametersGenerator.getSymbol(ctx.settings)
+        ctx.delegator.useFileWriter(paramsSymbol) {
+            EndpointParametersGenerator(it, rules).render()
+        }
+    }
+
+    /**
+     * Generate an implementation and supporting code for the modeled endpoint provider.
```
Comment (lauzadis) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointDelegator.kt:56]: This just generates the test cases right? Not the implementation and support code for the endpoint provider?

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt [hunk 2]
```diff
@@ -260,7 +272,12 @@ internal class HttpAuthHandler<Input, Output>(
         val identityProvider = authScheme.identityProvider(authConfig.identityProviderConfig)
         val identity = identityProvider.resolve(authOption.attributes)
 
-        // FIXME - endpoint resolution should happen here, either we make it explicit or we change implementation to use modifyBeforeSigning and have to stuff identity into the context
+        val resolveEndpointReq = ResolveEndpointRequest(request.context, request.subject.immutableView(), identity)
+        val endpoint = endpointResolver?.resolve(resolveEndpointReq)
+        if (endpoint != null) {
+            coroutineContext.debug<AuthHandler<*, *>> { "resolved endpoint: $endpoint" }
```
Comment (lauzadis) [inline runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt:278]: question: Why not `AuthHandler<Input, Output>`?

**Review by lauzadis — APPROVED** [review]

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt (hunk 2 above)
Comment (aajtodd) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt:278]: The `reified` type parameter isn't used for anything other than the `qualified` class name which doesn't retain generic type info anyway ([example](https://pl.kotl.in/Oq49gl2ln))

**Review by lucix-aws — APPROVED** [review]

**Review by ianbotsf — APPROVED** [review]

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/model/knowledge/EndpointParameterIndex.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+     *
+     * @param op operation to check parameters for
+     * @return true if there are any static or input context parameters for the given operation
+     */
+    fun hasContextParams(op: OperationShape): Boolean =
+        staticContextParams(op) != null || inputContextParams(op).isNotEmpty()
+
+    companion object {
+        fun of(model: Model): EndpointParameterIndex = EndpointParameterIndex(model)
+    }
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/model/knowledge/EndpointParameterIndex.kt:61]: It's not necessary but it does follow the same convention that all the smithy provided knowledge indexes use. I figured we'd keep the knowledge indexes consistent but I don't have a strong preference.

**Pull request merged**